xmovie/core.py

```diff
@@ -358,12 +370,63 @@ def save_frames(self, odir, progress=False):
         elif ~tqdm_avail and progress:
             warnings.warn("Cant show progess bar at this point. Install tqdm")
 
-        for fi in frame_range:
-            fig, ax, pp = self.render_frame(fi)
-            frame_save(
-                fig, fi, odir=odir, frame_pattern=self.frame_pattern, dpi=self.dpi
+        for timestep in frame_range:
+            fig, ax, pp = self.render_single_frame(timestep)
+            save_single_frame(
+                fig, timestep, odir=odir, frame_pattern=self.frame_pattern, dpi=self.dpi
             )
 
+
+
+    def save_frames_parallel(self, odir):
+        '''
+        Saves all frames in parallel using dask.map_blocks.
+
+        Parameters
+        ----------
+        odir : path
+            path to output directory
+        '''
+        import numpy as np
+        import dask.array as darray
+
+        da = self.data
+        framedim = self.framedim
+        
+        assert isinstance(da, xr.DataArray)
+        
+        # move 'framedim' to be last dimension if not already
+        if da.dims[-1] != framedim:
+            new_dims = list(da.dims)
+            new_dims.append(new_dims.pop(new_dims.index(framedim))) # Move framedim to last axis
+            da = da.transpose(*new_dims)
+
+        da = da.chunk({self.framedim : 1}) # DataArray needs to have length-1 chunks in framedim
+        chunk_dims = [ dim for dim in da.dims if dim != framedim ] # Create dimensions for each chunk
+                
+        def create_render_save_single_frame(xr_array, timestep):
+            fig, ax, pp = self.render_single_frame(timestep)
+            save_single_frame(
+                fig, timestep, odir=odir, frame_pattern=self.frame_pattern, dpi=self.dpi
+            )
+            return
+
+
+        mapped_save_and_close_frames = darray.map_blocks(_chunk_wrapper,
```

Comment:
use `xarray.map_blocks` instead.

Reply:
I think I can understand why that is a good idea: no need for a wrapper function etc. However I don't know how to pass the information of which chunk each processor has to the plotting function. In the wrapper example this is done using `block_info`, but apparently `xarray.map_block` doesn't have that feature.

Unless I'm missing something here, there's no way to use `xarray.map_blocks` without some wrapper function if I can't get info on the block, no? Which makes it seem to me that there's no advantage to `xarray.map_blocks`.

Here's the sort of stuff I've tried:

```python
        def create_render_save_single_frame2(xr_array, block_info=None):
            timestep = block_info[None]['chunk-location'][-1]
            fig, ax, pp = self.render_single_frame(timestep)
            save_single_frame(
                fig, timestep, odir=odir, frame_pattern=self.frame_pattern, dpi=self.dpi
            )
            return xr_array


        mapped_save_and_close_frames = da.map_blocks(func=create_render_save_single_frame2,
                                                     template=da,
                                                     )
        mapped_save_and_close_frames.compute()
```

Which of course fails with `TypeError: 'NoneType' object is not subscriptable` since `xarray` doesn't pass the `block_info` variable.

Reply:
the arrays coordinate variables tell you which "block" you have.

Reply:
> the arrays coordinate variables tell you which "block" you have.

Thanks! That's a very logical and simple answer and I'm ashamed I didn't realize this earlier. That's the whole point of using `xarray`. 

Also changed the return value to `timestep` like you suggested.

I've been trying this but since I can't print things inside the mapped function it's pretty hard for me to debug. I'm not familiar at all with dask so some of these errors are pretty obscure to me. I'm pushing this as it is for now in case you wanna take a quick crack at it. I'm sure you can solve whatever's happening in 5 min. Just FYI, at the moment the error I'm getting is: `ValueError: some chunks keys are not dimensions on this object: {'time'}`

Reply:
I just realized that you need access to my repo to try to modify this PR. Let me know if you'd like to do that.

It would definitely help me since I don't have much more time to dedicate to this (I gotta get back to research!), so I might need to throw in the towel soon and let the PR go stale if I can't get it to work...

Reply:
> I can't print things inside the mapped function it's pretty hard for me to debug. 

Use `.compute(scheduler="single-threaded")` this is equivalent to running a for loop over the dask blocks, you can use your standard debugging tricks then :)

> It would definitely help me since I don't have much more time to dedicate to this (I gotta get back to research!)

but isn't this more interesting? :P

Reply:
> Use .compute(scheduler="single-threaded") this is equivalent to running a for loop over the dask blocks, you can use your standard debugging tricks then :)

Awesome tip. Thanks!

> but isn't this more interesting? :P

Indeed. But if I were to always do the most interesting thing all the time I probably wouldn't have a job lol